### ISSUE TITLE
Networks: How to define ResNet-like Skip Connections?

### ISSUE THREAD

**Issue by Luca96** [tensorforce/tensorforce#673]
Hi @AlexKuhnle, I'm struggling to add some sort of skip connections to my policy network... Could you help me? 
Let's say I have a network with three inputs, in which each input is processed in isolation meaning that the network is made of three sub-networks like this one:
```bash
Inputs:  [A, B, C]
N1: A -> conv -> conv -> pool -> A_out
N2: B -> dense -> dense -> B_out
N3: C -> dense -> dense -> C_out
Output: concat [ A_out, B_out, C_out ]
```
Actually, I'm able to define such kind of multi-input networks but what if I want to reuse the output of some layer in the middle of a (sub)network? I tried to do the above, unfortunately without success:
```bash
[ retrieve "A",
  reshape,
  register "reshape_out",  # my intention: keep reshape's output
  conv2d,
  conv2d,
  pool,
  register "conv_out",  # my intention: keep pool's output
  retrieve "reshape_out" 
  some_layer       # my intention: apply 'some_layer' to "reshape_out" not 'conv_out'
  reshape
  register "some_layer_out"
  retrieve ["conv_out", "some_layer_out"] aggregation='product'
  register "A_out"
]
```
The above crashes [here](https://github.com/tensorforce/tensorforce/blob/master/tensorforce/core/layers/misc.py#L336). The error message states that the shapes don't match.

Stacktrace:
```bash
File "/home/luca/git/tensorforce/tensorforce/agents/agent.py", line 116, in create
    return Agent.create(agent=agent, environment=environment, **kwargs)
  File "/home/luca/git/tensorforce/tensorforce/agents/agent.py", line 87, in create
    agent = agent(**kwargs)
  File "/home/luca/git/tensorforce/tensorforce/agents/ppo.py", line 278, in __init__
    super().__init__(
  File "/home/luca/git/tensorforce/tensorforce/agents/tensorforce.py", line 313, in __init__
    self.model = TensorforceModel(
  File "/home/luca/git/tensorforce/tensorforce/core/models/tensorforce.py", line 102, in __init__
    self.policy = self.add_module(
  File "/home/luca/git/tensorforce/tensorforce/core/module.py", line 1186, in add_module
    module = module_cls(name, **kwargs)
  File "/home/luca/git/tensorforce/tensorforce/core/policies/parametrized_distributions.py", line 77, in __init__
    self.network = self.add_module(
  File "/home/luca/git/tensorforce/tensorforce/core/module.py", line 1188, in add_module
    module = module_cls(name, first_arg, **kwargs)
  File "/home/luca/git/tensorforce/tensorforce/core/networks/layered.py", line 58, in __init__
    self.parse_layers_spec(layers_spec=self.layers_spec, layer_counter=Counter())
  File "/home/luca/git/tensorforce/tensorforce/core/networks/layered.py", line 63, in parse_layers_spec
    self.parse_layers_spec(layers_spec=spec, layer_counter=layer_counter)
  File "/home/luca/git/tensorforce/tensorforce/core/networks/layered.py", line 63, in parse_layers_spec
    self.parse_layers_spec(layers_spec=spec, layer_counter=layer_counter)
  File "/home/luca/git/tensorforce/tensorforce/core/networks/layered.py", line 78, in parse_layers_spec
    self.add_module(name=layer_name, module=layers_spec)
  File "/home/luca/git/tensorforce/tensorforce/core/networks/network.py", line 192, in add_module
    layer = super().add_module(*args, **kwargs)
  File "/home/luca/git/tensorforce/tensorforce/core/module.py", line 1186, in add_module
    module = module_cls(name, **kwargs)
  File "/home/luca/git/tensorforce/tensorforce/core/layers/misc.py", line 329, in __init__
    super().__init__(name=name, input_spec=input_spec, summary_labels=summary_labels)
  File "/home/luca/git/tensorforce/tensorforce/core/layers/layer.py", line 63, in __init__
    self.output_spec = self.get_output_spec(input_spec=dict(self.input_spec))
  File "/home/luca/git/tensorforce/tensorforce/core/layers/misc.py", line 336, in get_output_spec
    raise TensorforceError.value(name='Reshape', argument='shape', value=self.shape)
tensorforce.exception.TensorforceError: Invalid value for Reshape argument shape: 50.
```
The error refers to a `reshape` layer, which I think is the last one in the above example. `50` is the shape I want to reshape to, but it crashes because the reshape layer compares the shape `50` with the shape of the input (in my case `(50,40,1)`) and not the output's shape of `some_layer` (which, in my case is a `conv2d` layer).
Do you have any clue about that?

**Comment by AlexKuhnle** [comment]
Hey,

I can't see anything obviously wrong, could you post the actual specification so I can get a better idea about shapes (that's where the mismatch seems to be)? One thing that is confusing is why you have to use so many reshape layers, particularly the one at the very beginning? (Also, sure you want to use `product` aggregation and not `sum`, given you want to implement a residual layer?)

**Comment by Luca96** [comment]
Consider the following toy example:
```python
# State space:
dict(a=dict(type='float', shape=(10, 10)),
     b=dict(type='float', shape=(10, 17)))

# Action space:
dict(type='float', shape=(3,), min_value=-1.0, max_value=1.0)
```
This is my network specification (note: I just want to understand the mechanism in general, so I'm not limited to residual layers or similar):
```python
[
  # network A:
  [{
    'type': 'retrieve',
    'tensors': ['a'],
    'aggregation': 'concat'
  }, {
    'type': 'keras',
    'layer': 'LayerNormalization'
  }, {
    'type': 'reshape',
    'shape': (10, 10, 1)
  }, {
    'type': 'register',
    'tensor': 'reshape_out'
  }, {
    'type': 'conv2d',
    'size': 6,
    'window': (3, 3),
    'stride': 1,
    'padding': 'same',
    'bias': True,
    'activation': 'swish',
    'dropout': 0.0
  }, {
    'type': 'conv2d',
    'size': 12,
    'window': (3, 3),
    'stride': 1,
    'padding': 'same',
    'bias': True,
    'activation': 'swish',
    'dropout': 0.0
  }, {
    'type': 'conv2d',
    'size': 18,
    'window': (3, 3),
    'stride': 1,
    'padding': 'same',
    'bias': True,
    'activation': 'swish',
    'dropout': 0.0
  }, {
    'type': 'conv2d',
    'size': 24,
    'window': (3, 3),
    'stride': 1,
    'padding': 'same',
    'bias': True,
    'activation': 'swish',
    'dropout': 0.0
  }, {
    'type': 'pooling',
    'reduction': 'mean'
  }, {
    'type': 'register',
    'tensor': 'conv_out'
  }, {
    'type': 'retrieve',
    'tensors': ['reshape_out']
  }, {
    'type': 'conv2d',
    'size': 1,
    'window': (1, 10),
    'stride': 1,
    'activation': 'swish'
  }, {
    'type': 'reshape',
    'shape': (10, )
  }, {
    'type': 'register',
    'tensor': 'avg_out'
  }, {
    'type': 'retrieve',
    'tensors': ['conv_out', 'avg_out'],
    'aggregation': 'product'
  }, {
    'type': 'register',
    'tensor': 'a_out'
  }],
  # Network B:
  [{
    'type': 'retrieve',
    'tensors': ['b'],
    'aggregation': 'concat'
  }, {
    'type': 'keras',
    'layer': 'LayerNormalization'
  }, {
    'type': 'reshape',
    'shape': (10, 17, 1)
  }, {
    'type': 'register',
    'tensor': 'reshape_out'
  }, {
    'type': 'conv2d',
    'size': 6,
    'window': (3, 4),
    'stride': 1,
    'padding': 'same',
    'bias': True,
    'activation': 'swish',
    'dropout': 0.0
  }, {
    'type': 'conv2d',
    'size': 12,
    'window': (3, 4),
    'stride': 1,
    'padding': 'same',
    'bias': True,
    'activation': 'swish',
    'dropout': 0.0
  }, {
    'type': 'conv2d',
    'size': 18,
    'window': (3, 4),
    'stride': 1,
    'padding': 'same',
    'bias': True,
    'activation': 'swish',
    'dropout': 0.0
  }, {
    'type': 'conv2d',
    'size': 24,
    'window': (3, 4),
    'stride': 1,
    'padding': 'same',
    'bias': True,
    'activation': 'swish',
    'dropout': 0.0
  }, {
    'type': 'pooling',
    'reduction': 'mean'
  }, {
    'type': 'register',
    'tensor': 'conv_out'
  }, {
    'type': 'retrieve',
    'tensors': ['reshape_out']
  }, {
    'type': 'conv2d',
    'size': 1,
    'window': (1, 17),
    'stride': 1,
    'activation': 'swish'
  }, {
    'type': 'reshape',
    'shape': (10, )
  }, {
    'type': 'register',
    'tensor': 'avg_out'
  }, {
    'type': 'retrieve',
    'tensors': ['conv_out', 'avg_out'],
    'aggregation': 'product'
  }, {
    'type': 'register',
    'tensor': 'b_out'
  }], 
  # Concatenation of Network A and B output:
  {
    'type': 'retrieve',
    'tensors': ['a_out', 'b_out'],
    'aggregation': 'concat'
  }, {
    'type': 'dense',
    'size': 64,
    'activation': 'relu',
    'dropout': 0.0
  }, {
    'type': 'dense',
    'size': 64,
    'activation': 'relu',
    'dropout': 0.0
  }
]
```
In this case the error message is:
```bash
  File "/home/luca/git/tensorforce/tensorforce/core/layers/misc.py", line 337, in get_output_spec
    raise TensorforceError.value(name='Reshape', argument='shape', value=self.shape)
tensorforce.exception.TensorforceError: Invalid value for Reshape argument shape: 10.
```
...The stacktace is similar to the one in the above comment.
Hope it helps...

**Comment by AlexKuhnle** [comment]
Two problems with this network config: First, the default value for `conv2d` `padding` is `'same'`, which means that the window of `(1, 10/17)` will not reduce the output size to 1 --> you need to set `padding='valid'` instead. Second, `conv_out` has a shape of `(-1, 24)` (`pooling` collapses all but the last dimension), whereas `avg_out` has a shape of `(-1, 10)`, so the product aggregation is invalid. Minor comments: The initial `retrieve` doesn't need an `aggregation` value, it's ignored anyway. And the names for the tensors are re-used -- doesn't cause a problem with the current version, I think, but the next version likely won't allow this anymore. :-)

**Comment by jerabaul29** [comment]
That reminds me of discussions we had had on the theme of 'would it be possible to use a keras-like interface to describe complex ANNs for use in tensorforce'?

**Comment by AlexKuhnle** [comment]
You're right, this feature request has been around for a while... and with the upcoming changes to properly adopt TF2 this should not require much effort. One question (may have asked before): Does Keras offer a JSON-style specification format, or do you mean just something like a `KerasNetwork` class which makes it easily possible to use Keras models? (JSON-format would certainly be helpful.)

**Comment by Luca96** [comment]
@AlexKuhnle I found this [to_json()](https://www.tensorflow.org/api_docs/python/tf/keras/Model#to_json) method in the `tf.keras` docs which states that "Returns a JSON string containing the network configuration". So, at least is possibile to get a JSON description of Keras networks.

Anyway, I have some working code to abstract and simplify networks specifications. It looks like this:
```python
network = NetworkSpec(inputs=['A', 'B', 'C'], output='net_out')  # concat tensors A, B, and C

for _ in range(num_layers):
   network.conv2d(....)
   network.batch_normalization()

network.flatten()
network.dense(...)
spec = network.build()  # -> returns the network spec as a list of dict.
```
I plan to extend it to non-linear architectures as well, and send a PR when it's full working.

**Comment by jerabaul29** [comment]
I think that keras has nice, non JSON ways to define ANN architectures, that may be easier to use. See for example:

https://machinelearningmastery.com/keras-functional-api-deep-learning/

So as an end user I would think having an interface that uses / is similar to their Model would be great, if possible, and may give access to many of the Keras built-in tools (like plotting networks, etc). 

Models in Keras can be exported to json:

https://keras.io/models/about-keras-models/

So it may be possible to have the best of both worlds, defining easily models in Keras, doing inspection etc, and then exporting to JSON for use by tf?

**Comment by jerabaul29** [comment]
(ooops, some things are duplicate of the post by @Luca96 , sorry I think we were writing at the same time :) ).

**Comment by Luca96** [comment]
I found a nice [model_from_json](https://www.tensorflow.org/api_docs/python/tf/keras/models/model_from_json) API. ;) More details on saving and loading keras models, [here](https://www.tensorflow.org/guide/keras/save_and_serialize).

**Comment by AlexKuhnle** [comment]
@Luca96, looks like a fun module. It's basically a stand-alone utility, right?

@both: Non-JSON specification is of course fine and also possible right now, but it has implications on whether Tensorforce can store the configuration. And there are probably limits to what degree features of Keras `Model` (summaries, plotting, etc) are compatible with Tensorforce -- at least there are a lot of restrictions until now. Maybe the full TF2-version will get rid of most/all of them, but not sure right now.

**Comment by Luca96** [comment]
Here is a simple example:
```python
# 1. Model definition:
def build_model(input_shape):
    inputs = Input(shape=input_shape)
    x = Dense(64, activation='relu')(inputs)
    outputs = Dense(3, activation='softmax')(x)
    
    return Model(inputs, outputs, name='my_model')

# 2. Make the model:
model = build_model(16)
model.summary()

# Summary output:
Model: "my_model"
_________________________________________________________________
Model: "my_model"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
input_4 (InputLayer)         [(None, 16)]              0         
_________________________________________________________________
dense_8 (Dense)              (None, 64)                1088      
_________________________________________________________________
dense_9 (Dense)              (None, 3)                 195       
=================================================================
Total params: 1,283
Trainable params: 1,283
Non-trainable params: 0
_________________________________________________________________

# 3. Convert model to JSON:
json = model.to_json()
```
```json
# model.to_json() output:
{
  "class_name": "Model",
  "config": {
    "name": "my_model",
    "layers": [
      {
        "class_name": "InputLayer",
        "config": {
          "batch_input_shape": [
            null,
            16
          ],
          "dtype": "float32",
          "sparse": false,
          "ragged": false,
          "name": "input_5"
        },
        "name": "input_5",
        "inbound_nodes": []
      },
      {
        "class_name": "Dense",
        "config": {
          "name": "dense_10",
          "trainable": true,
          "dtype": "float32",
          "units": 64,
          "activation": "relu",
          "use_bias": true,
          "kernel_initializer": {
            "class_name": "GlorotUniform",
            "config": {
              "seed": null
            }
          },
          "bias_initializer": {
            "class_name": "Zeros",
            "config": {}
          },
          "kernel_regularizer": null,
          "bias_regularizer": null,
          "activity_regularizer": null,
          "kernel_constraint": null,
          "bias_constraint": null
        },
        "name": "dense_10",
        "inbound_nodes": [
          [
            [
              "input_5",
              0,
              0,
              {}
            ]
          ]
        ]
      },
      {
        "class_name": "Dense",
        "config": {
          "name": "dense_11",
          "trainable": true,
          "dtype": "float32",
          "units": 3,
          "activation": "softmax",
          "use_bias": true,
          "kernel_initializer": {
            "class_name": "GlorotUniform",
            "config": {
              "seed": null
            }
          },
          "bias_initializer": {
            "class_name": "Zeros",
            "config": {}
          },
          "kernel_regularizer": null,
          "bias_regularizer": null,
          "activity_regularizer": null,
          "kernel_constraint": null,
          "bias_constraint": null
        },
        "name": "dense_11",
        "inbound_nodes": [
          [
            [
              "dense_10",
              0,
              0,
              {}
            ]
          ]
        ]
      }
    ],
    "input_layers": [
      [
        "input_5",
        0,
        0
      ]
    ],
    "output_layers": [
      [
        "dense_11",
        0,
        0
      ]
    ]
  },
  "keras_version": "2.2.4-tf",
  "backend": "tensorflow"
}
```
```python
# 4. Loading the model from JSON:
loaded_model = tf.keras.models.model_from_json(json)
loaded_model.summary()  # same output as before
```

That's it. :D

**Comment by AlexKuhnle** [comment]
Looks great. I think the best way is to implement it as a separate network class (`network=dict(type='keras', spec={...})`), and/or maybe there could be some auto-detection. Loading would be good as well -- I hope the new TF2 saving/loading mechanism makes it easier to integrate this, the old `Saver` is not super straightforward.

**Comment by Luca96** [comment]
@AlexKuhnle, do you remember https://github.com/tensorforce/tensorforce/issues/673#issuecomment-622960169? The changes you suggested here https://github.com/tensorforce/tensorforce/issues/673#issuecomment-623184952 are not enough to solve the issue with the network's architecture...
1. Actually, having registered more than one tensor with the **same name** (e.g. `reshape_out` for both network `A` and `B` input tensors) causes a failure. It seems that input tensors got confused, so shapes don't match anymore. Here is the exception message:
```bash
File "/home/luca/git/tensorforce/tensorforce/agents/agent.py", line 116, in create
    return Agent.create(agent=agent, environment=environment, **kwargs)
  File "/home/luca/git/tensorforce/tensorforce/agents/agent.py", line 87, in create
    agent = agent(**kwargs)
  File "/home/luca/git/tensorforce/tensorforce/agents/ppo.py", line 281, in __init__
    super().__init__(
  File "/home/luca/git/tensorforce/tensorforce/agents/tensorforce.py", line 313, in __init__
    self.model = TensorforceModel(
  File "/home/luca/git/tensorforce/tensorforce/core/models/tensorforce.py", line 102, in __init__
    self.policy = self.add_module(
  File "/home/luca/git/tensorforce/tensorforce/core/module.py", line 1186, in add_module
    module = module_cls(name, **kwargs)
  File "/home/luca/git/tensorforce/tensorforce/core/policies/parametrized_distributions.py", line 83, in __init__
    self.network = self.add_module(
  File "/home/luca/git/tensorforce/tensorforce/core/module.py", line 1188, in add_module
    module = module_cls(name, first_arg, **kwargs)
  File "/home/luca/git/tensorforce/tensorforce/core/networks/layered.py", line 58, in __init__
    self.parse_layers_spec(layers_spec=self.layers_spec, layer_counter=Counter())
  File "/home/luca/git/tensorforce/tensorforce/core/networks/layered.py", line 63, in parse_layers_spec
    self.parse_layers_spec(layers_spec=spec, layer_counter=layer_counter)
  File "/home/luca/git/tensorforce/tensorforce/core/networks/layered.py", line 63, in parse_layers_spec
    self.parse_layers_spec(layers_spec=spec, layer_counter=layer_counter)
  File "/home/luca/git/tensorforce/tensorforce/core/networks/layered.py", line 78, in parse_layers_spec
    self.add_module(name=layer_name, module=layers_spec)
  File "/home/luca/git/tensorforce/tensorforce/core/networks/network.py", line 192, in add_module
    layer = super().add_module(*args, **kwargs)
  File "/home/luca/git/tensorforce/tensorforce/core/module.py", line 1186, in add_module
    module = module_cls(name, **kwargs)
  File "/home/luca/git/tensorforce/tensorforce/core/layers/misc.py", line 293, in __init__
    Module.register_tensor(name=self.tensor, spec=self.input_spec, batched=True)
  File "/home/luca/git/tensorforce/tensorforce/core/module.py", line 76, in register_tensor
    raise TensorforceError.mismatch(
tensorforce.exception.TensorforceError: Module.register_tensor mismatch for argument spec: {'type': 'float', 'shape': (10, 17, 1), 'batched': True} <-> {'type': 'float', 'shape': (10, 10, 1), 'batched': True}.
```
2. Another weird thing: after registering the first reshape layer, I had to retrieve it soon to make the coming `conv2d` have the right shapes. That's the situation:
```python
[ ...
  reshape
  register reshape
  retrieve reshape  # !?
  conv2d
 ...
]
```
This works. But if I remove the `retrieve reshape`, then the conv shapes got negative being `(-2, -2, 6)`. Note: the input shape is `(10, 10, 1)` and this happens at the first conv layer...
I put a print statement right after [this line](https://github.com/tensorforce/tensorforce/blob/master/tensorforce/core/layers/layer.py#L63) and got this:
```python
keras0 {'type': 'float', 'shape': (10, 10)} {'type': 'float', 'shape': (10, 10)}
reshape0 {'type': 'float', 'shape': (10, 10)} {'type': 'float', 'shape': (10, 10, 1)}
register0 {'type': 'float', 'shape': (10, 10, 1)} {'type': 'float', 'shape': (10, 10, 1)}
conv2d0 {'type': 'float', 'shape': (0, 0, 0)} {'type': 'float', 'shape': (-2, -2, 6)}  # <- negative shapes
# crash
```
It seems that the `conv` layer got some wrong input shapes: `(0,0,0)` instead of `(10,10,1)`.

The error message is:
```bash
File "/home/luca/PycharmProjects/carla-driving-rl-agent/tests.py", line 198, in toy_example
    agent = Agent.create(agent='ppo',
  File "/home/luca/git/tensorforce/tensorforce/agents/agent.py", line 116, in create
    return Agent.create(agent=agent, environment=environment, **kwargs)
  File "/home/luca/git/tensorforce/tensorforce/agents/agent.py", line 87, in create
    agent = agent(**kwargs)
  File "/home/luca/git/tensorforce/tensorforce/agents/ppo.py", line 281, in __init__
    super().__init__(
  File "/home/luca/git/tensorforce/tensorforce/agents/tensorforce.py", line 313, in __init__
    self.model = TensorforceModel(
  File "/home/luca/git/tensorforce/tensorforce/core/models/tensorforce.py", line 102, in __init__
    self.policy = self.add_module(
  File "/home/luca/git/tensorforce/tensorforce/core/module.py", line 1186, in add_module
    module = module_cls(name, **kwargs)
  File "/home/luca/git/tensorforce/tensorforce/core/policies/parametrized_distributions.py", line 83, in __init__
    self.network = self.add_module(
  File "/home/luca/git/tensorforce/tensorforce/core/module.py", line 1188, in add_module
keras0 {'type': 'float', 'shape': (10, 10)} {'type': 'float', 'shape': (10, 10)}
reshape0 {'type': 'float', 'shape': (10, 10)} {'type': 'float', 'shape': (10, 10, 1)}
register0 {'type': 'float', 'shape': (10, 10, 1)} {'type': 'float', 'shape': (10, 10, 1)}
conv2d0 {'type': 'float', 'shape': (0, 0, 0)} {'type': 'float', 'shape': (-2, -2, 6)}
    module = module_cls(name, first_arg, **kwargs)
  File "/home/luca/git/tensorforce/tensorforce/core/networks/layered.py", line 58, in __init__
    self.parse_layers_spec(layers_spec=self.layers_spec, layer_counter=Counter())
  File "/home/luca/git/tensorforce/tensorforce/core/networks/layered.py", line 63, in parse_layers_spec
    self.parse_layers_spec(layers_spec=spec, layer_counter=layer_counter)
  File "/home/luca/git/tensorforce/tensorforce/core/networks/layered.py", line 63, in parse_layers_spec
    self.parse_layers_spec(layers_spec=spec, layer_counter=layer_counter)
  File "/home/luca/git/tensorforce/tensorforce/core/networks/layered.py", line 78, in parse_layers_spec
    self.add_module(name=layer_name, module=layers_spec)
  File "/home/luca/git/tensorforce/tensorforce/core/networks/network.py", line 192, in add_module
    layer = super().add_module(*args, **kwargs)
  File "/home/luca/git/tensorforce/tensorforce/core/module.py", line 1186, in add_module
    module = module_cls(name, **kwargs)
  File "/home/luca/git/tensorforce/tensorforce/core/layers/convolution.py", line 182, in __init__
    super().__init__(
  File "/home/luca/git/tensorforce/tensorforce/core/layers/layer.py", line 170, in __init__
    super().__init__(
  File "/home/luca/git/tensorforce/tensorforce/core/layers/layer.py", line 65, in __init__
    self.output_spec = util.valid_value_spec(
  File "/home/luca/git/tensorforce/tensorforce/util.py", line 461, in valid_value_spec
    raise TensorforceError.value(name=value_type, argument='shape', value=shape)
tensorforce.exception.TensorforceError: Invalid value for tensor argument shape: -2,-2,6.
```
To sum up, I managed to define such network architecture but I had to: (1) use different names for registered tensors, and (2) add another `retrieve "reshape"` layer right after its `register`. :thinking:

**Comment by AlexKuhnle** [comment]
That the tensor names have to be different makes sense -- forgot that obviously, if their shape differs, the registry will complain. The other problem doesn't make sense, so there may probably be a bug somewhere...

**Comment by AlexKuhnle** [comment]
The internal mechanisms have now changed, so best to open a new issue if new problems come up.
